File: src/main/java/org/mitallast/queue/common/netty/NettyClient.java
package org.mitallast.queue.common.netty;

import com.typesafe.config.Config;
import io.netty.channel.Channel;
import io.netty.channel.ChannelInitializer;

public abstract class NettyClient extends NettyClientBootstrap {

    protected final String host;
    protected final int port;
    protected volatile Channel channel;

    protected NettyClient(Config config, NettyProvider provider, String host, int port) {
        super(config, provider);
        this.host = host;
        this.port = port;
    }

    @Override
    protected void doStart() {
        super.doStart();
        try {
            channel = connect(host, port).sync().channel();
            init();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    protected void init() {
    }

    @Override
    protected void doStop() {
        channel.close().awaitUninterruptibly();
        super.doStop();
    }

    protected abstract ChannelInitializer channelInitializer();
}

File: src/main/java/org/mitallast/queue/common/netty/NettyClient.kt
package org.mitallast.queue.common.netty

import com.typesafe.config.Config
import io.netty.channel.Channel
import io.netty.channel.ChannelInitializer
import org.mitallast.queue.common.logging.LoggingService

abstract class NettyClient protected constructor(
    config: Config,
    logging: LoggingService,
    provider: NettyProvider,
    protected val host: String,
    protected val port: Int
) : NettyClientBootstrap(config, logging, provider) {
    @Volatile
    protected var channel: Channel? = null

    init {
        this.channel = null
    }

    override fun doStart() {
        super.doStart()
        try {
            channel = connect(host, port).sync().channel()
            init()
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }

    }

    protected fun init() {}

    override fun doStop() {
        if (channel != null) {
            channel!!.close().awaitUninterruptibly()
            channel = null
        }
        super.doStop()
    }

    abstract override fun channelInitializer(): ChannelInitializer<*>
}
